File: easy/94_binary_tree_inorder_traversal/main.rb
#
# @lc app=leetcode id=94 lang=ruby
#
# [94] Binary Tree Inorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  @path = []
  dfs(root)
  @path
end

def dfs(node)
  return if node.nil?

  dfs(node.left)
  @path << node.val
  dfs(node.right)
end
# @lc code=end

root = TreeNode.new(1, nil, TreeNode.new(2, TreeNode.new(3)))
p inorder_traversal(root)

File: easy/94_binary_tree_inorder_traversal/main.go
package main

import "fmt"

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	path := []int{}
	dfs(root, &path)
	return path
}

func dfs(node *TreeNode, path *[]int) {
	if node == nil {
		return
	}

	dfs(node.Left, path)
	*path = append(*path, node.Val)
	dfs(node.Right, path)
}

func main() {
	// root = [1,null,2,3]
	root := &TreeNode{
		Val: 1,
		Right: &TreeNode{
			Val:  2,
			Left: &TreeNode{Val: 3}}}
	fmt.Println(inorderTraversal(root))
}
